### ISSUE TITLE
Using built-in http.Client instead fasthttp.Client

### ISSUE THREAD

**Issue by Ja7ad** [meilisearch/meilisearch-go#554]
The fasthttp library isn't very appealing as a client and is mostly used on the server side. However, its only advantage is the use of sync.Pool to reuse an object and improve memory usage.

But as an SDK client, we need to try to use built-in packages to avoid package conflicts for users during usage. Most major SDKs like AWS and Google Cloud use the built-in HTTP client package.

- https://github.com/aws/aws-sdk-go
- https://github.com/googleapis/google-cloud-go

However, regarding context.Context, we need to receive a higher-level stop signal so that the request can be halted.

#462

There are many solutions to improve the performance of http.Client, and we can enhance it quite easily.

We are currently unstable and we can improve our sdk 🚀, I can refactor it for stable.

**Comment by curquiza** [comment]
Hello @Ja7ad 

Thank you for this suggestion. You got a point!

How big is the refactoring? I mean, how many PRs? If you want we can create a temporary branch to make you merge smaller PRs on it (and avoid having a stale states on `main`)

If we do this, I need to clearly see what is breaking
- what the users did before
- what they have to do instead with the new version

I really thank you for your involvement in this repository 🙏

**Comment by Ja7ad** [comment]
> Hello @Ja7ad 
> 
> Thank you for this suggestion. You got a point!
> 
> How big is the refactoring? I mean, how many PRs? If you want we can create a temporary branch to make you merge smaller PRs on it (and avoid having a stale states on `main`)
> 
> If we do this, I need to clearly see what is breaking
> - what the users did before
> - what they have to do instead with the new version
> 
> I really thank you for your involvement in this repository 🙏 

Please make temporary branch refactor

**Comment by curquiza** [comment]
@Ja7ad 
I created the `refactor` branch
You can merge your PRs to `refactor`

Be clear and transparent about breaking changes for the users.

> In this refactor all method have context.Context parameter and many changes on client.

Will this parameter mandatory?

Also, before starting anything: what is the purpose of this refacor? What is the benefit as a user?

**Comment by Ja7ad** [comment]
> @Ja7ad I created the `refactor` branch You can merge your PRs to `refactor`
> 
> Be clear and transparent about breaking changes for the users.
> 
> > In this refactor all method have context.Context parameter and many changes on client.
> 
> Will this parameter be mandatory?


Yes, But we can use standard approach with or without context.Context as mandatory.

similar `http.NewRequest` or `http.NewRequestWithContext`.

We can have 2 method for any API for example.

```go
resp, err := index.FetchInfo() // This is wrapper of FetchInfoWithContext just pass inputs to FetchInfoWithContext with context.Background()

resp, err := index.FetchInfoWithContext(ctx context.Context)
```

> Also, before starting anything: what is the purpose of this refacor? What is the benefit as a user?

- Refactor client with http.Client with optimized configuration
- Support parent context for all methods
- Add option pattern for optional configuration of constructor
- Improved documentation package, methods

**Comment by madlitz** [comment]
@curquiza 

> What is the benefit as a user?

I just landed here after evaluating Typesense vs Meilisearch for our Golang backend. It was indeed a red flag when I looked at the Go SDK API and saw methods did not take a context and were async by default. This is quite non-standard in the Go ecosystem.

Agree also that removing fasthttp as a dependency is a nice win.

Happy that this is being actively considered and I am willing to help with the refactor.

**Comment by Ja7ad** [comment]
> @curquiza
> 
> > What is the benefit as a user?
> 
> I just landed here after evaluating Typesense vs Meilisearch for our Golang backend. It was indeed a red flag when I looked at the Go SDK API and saw methods did not take a context and were async by default. This is quite non-standard in the Go ecosystem.
> 
> Agree also that removing fasthttp as a dependency is a nice win.
> 
> Happy that this is being actively considered and I am willing to help with the refactor.

Issue implemented on this PR: https://github.com/meilisearch/meilisearch-go/pull/556

**Comment by curquiza** [comment]
@Ja7ad I opened #557 if you need to merge other PRs

**Comment by Ja7ad** [comment]
> @Ja7ad I opened #557 if you need to merge other PRs

Hello thanks, I will add #540 #507 #501 #498 #488 tomorrow on main for new refactor.

Please merge refactor to main.

**Comment by curquiza** [comment]
Ok, so we are good with the refactor? Nothing more needed?
I could have merged #556 directly to main 😅 hahaha

**Comment by Ja7ad** [comment]
> Ok, so we are good with the refactor? Nothing more needed? I could have merged #556 directly to main 😅 hahaha

Sure, cool
